File: csharp/02-variables/Variables.cs
// ── Primitive Types ──────────────────────────────────────────────────────────
// Integers
int age = 30;
long population = 8_100_000_000L;   // underscores improve readability
short smallNum = 32000;
byte byteVal = 255;

// Floating-point
double pi = 3.14159265358979;
float approxPi = 3.14159f;          // f suffix required for float literals
decimal price = 19.99m;             // m suffix for decimal; use for money

// Boolean
bool isActive = true;

// Characters and strings
char grade = 'A';                   // single quotes for char
string name = "C# Developer";       // double quotes for string

Console.WriteLine("── Primitive Types ──");
Console.WriteLine($"int:     {age}");
Console.WriteLine($"long:    {population}");
Console.WriteLine($"short:   {smallNum}");
Console.WriteLine($"byte:    {byteVal}");
Console.WriteLine($"double:  {pi}");
Console.WriteLine($"float:   {approxPi}");
Console.WriteLine($"decimal: {price}");
Console.WriteLine($"bool:    {isActive}");
Console.WriteLine($"char:    {grade}");
Console.WriteLine($"string:  {name}");

// ── Type Inference with var ───────────────────────────────────────────────────
// var lets the compiler deduce the type; the variable is still statically typed
var count = 42;                    // inferred as int
var message = "Hello, C#!";       // inferred as string
var ratio = 3.14;                  // inferred as double
var flag = true;                   // inferred as bool

Console.WriteLine("\n── Type Inference with var ──");
Console.WriteLine($"count   is {count.GetType().Name}: {count}");
Console.WriteLine($"message is {message.GetType().Name}: {message}");
Console.WriteLine($"ratio   is {ratio.GetType().Name}: {ratio}");
Console.WriteLine($"flag    is {flag.GetType().Name}: {flag}");

// var works great with complex types to reduce verbosity
var numbers = new List<int> { 1, 2, 3, 4, 5 };
Console.WriteLine($"numbers is {numbers.GetType().Name} with {numbers.Count} elements");

// ── Constants and Nullable Types ─────────────────────────────────────────────
// const: value must be known at compile time
const double Pi = 3.14159265358979;
const int DaysInWeek = 7;
const string AppName = "CodeArchaeology";

double circumference = 2 * Pi * 5.0;
Console.WriteLine("\n── Constants ──");
Console.WriteLine($"Pi          = {Pi}");
Console.WriteLine($"DaysInWeek  = {DaysInWeek}");
Console.WriteLine($"AppName     = {AppName}");
Console.WriteLine($"Circumference of r=5 circle: {circumference:F4}");

// Nullable value types: add ? to allow null for value types
int? optionalAge = null;
Console.WriteLine($"\noptionalAge is null: {optionalAge == null}");
optionalAge = 25;
Console.WriteLine($"optionalAge after assignment: {optionalAge}");

// Null-coalescing operator: return left if not null, else right
int definiteAge = optionalAge ?? 0;
Console.WriteLine($"definiteAge (using ??): {definiteAge}");

// ── Type Conversions ─────────────────────────────────────────────────────────
// Implicit conversions (widening)
int intVal = 100;
long longVal = intVal;           // int → long: always safe
double doubleVal = intVal;       // int → double: always safe

Console.WriteLine("\n── Type Conversions ──");
Console.WriteLine("Implicit conversions:");
Console.WriteLine($"  int    → long:   {longVal}");
Console.WriteLine($"  int    → double: {doubleVal}");

// Explicit casting (narrowing)
double piApprox = 3.99;
int truncated = (int)piApprox;   // fractional part is discarded, not rounded

Console.WriteLine("\nExplicit cast (truncates, does not round):");
Console.WriteLine($"  (int)3.99 = {truncated}");

// Convert class: parses strings and converts between types
string numberText = "42";
int parsed = Convert.ToInt32(numberText);
double parsedDouble = Convert.ToDouble("3.14");
bool parsedBool = Convert.ToBoolean(1);   // 0 = false, non-zero = true

Console.WriteLine("\nConvert class:");
Console.WriteLine($"  Convert.ToInt32(\"42\")      = {parsed}");
Console.WriteLine($"  Convert.ToDouble(\"3.14\")   = {parsedDouble}");
Console.WriteLine($"  Convert.ToBoolean(1)       = {parsedBool}");

// int.Parse / int.TryParse
bool success = int.TryParse("123abc", out int result);
Console.WriteLine($"\nint.TryParse(\"123abc\") succeeded: {success}");

success = int.TryParse("999", out result);
Console.WriteLine($"int.TryParse(\"999\")    succeeded: {success}, value: {result}");

// String representations
int x = 255;
Console.WriteLine($"\nString representations of 255:");
Console.WriteLine($"  Decimal: {x.ToString()}");
Console.WriteLine($"  Hex:     {x:X}");
Console.WriteLine($"  Binary:  {Convert.ToString(x, 2)}");

File: dart/02-variables/variables.dart
void main() {
  // ── Primitive Types ──────────────────────────────────────────────────────
  // Integers (arbitrary precision)
  int age = 30;
  int population = 8100000000;

  // Floating-point (64-bit double)
  double pi = 3.14159265358979;
  double temperature = -40.5;

  // Booleans
  bool isActive = true;

  // Strings (single or double quotes)
  String name = 'Dart Developer';
  String greeting = "Hello";

  print('── Primitive Types ──');
  print('int:    $age');
  print('int:    $population');
  print('double: $pi');
  print('double: $temperature');
  print('bool:   $isActive');
  print('String: $name');
  print('String: $greeting');

  // ── Type Inference with var ──────────────────────────────────────────────
  // var lets the compiler deduce the type; the variable is still statically typed
  var count = 42;                    // inferred as int
  var message = 'Hello, Dart!';     // inferred as String
  var ratio = 3.14;                  // inferred as double
  var flag = true;                   // inferred as bool

  print('\n── Type Inference with var ──');
  print('count   is ${count.runtimeType}: $count');
  print('message is ${message.runtimeType}: $message');
  print('ratio   is ${ratio.runtimeType}: $ratio');
  print('flag    is ${flag.runtimeType}: $flag');

  // var works great with complex types to reduce verbosity
  var numbers = [1, 2, 3, 4, 5];
  print('numbers is ${numbers.runtimeType} with ${numbers.length} elements');

  // ── final and const ──────────────────────────────────────────────────────
  // final: set once at runtime, cannot be reassigned
  final String appName = 'CodeArchaeology';
  final now = DateTime.now();    // runtime value — only final works here

  // const: compile-time constant — value must be known at compile time
  const double piConst = 3.14159265358979;
  const int daysInWeek = 7;
  const maxRetries = 3;          // type inferred as int

  double circumference = 2 * piConst * 5.0;

  print('\n── final and const ──');
  print('final appName:  $appName');
  print('final now:      ${now.year}-${now.month.toString().padLeft(2, '0')}-${now.day.toString().padLeft(2, '0')}');
  print('const piConst:  $piConst');
  print('const daysInWeek: $daysInWeek');
  print('const maxRetries: $maxRetries');
  print('Circumference of r=5 circle: ${circumference.toStringAsFixed(4)}');

  // ── Null Safety ──────────────────────────────────────────────────────────
  // By default, variables cannot be null
  // Add ? to make a type nullable
  String? nullableName;            // defaults to null
  int? optionalAge;

  print('\n── Null Safety ──');
  print('nullableName is null: ${nullableName == null}');
  print('optionalAge is null:  ${optionalAge == null}');

  // Null-aware operators
  nullableName = 'Alice';
  print('nullableName after assignment: $nullableName');
  print('nullableName length: ${nullableName.length}');   // safe — compiler knows it's assigned

  String? maybeName;
  print('maybeName?.length: ${maybeName?.length}');       // null-aware access → null
  print('maybeName ?? "default": ${maybeName ?? "default"}');  // null-coalescing

  // ── Type Conversions ─────────────────────────────────────────────────────
  // Dart does NOT do implicit numeric conversions — you must be explicit
  int intVal = 100;
  double doubleVal = intVal.toDouble();   // int → double
  int backToInt = 3.99.toInt();           // double → int (truncates)

  print('\n── Type Conversions ──');
  print('int → double: $intVal → $doubleVal');
  print('double → int (truncates): 3.99 → $backToInt');

  // String parsing
  int parsed = int.parse('42');
  double parsedDouble = double.parse('3.14');
  bool parsedBool = bool.parse('true');   // only 'true' or 'false'

  print('\nString parsing:');
  print('  int.parse("42")      = $parsed');
  print('  double.parse("3.14") = $parsedDouble');
  print('  bool.parse("true")   = $parsedBool');

  // Safe parsing with tryParse (returns null on failure)
  int? safeParse = int.tryParse('123abc');
  print('\nint.tryParse("123abc") = $safeParse');

  int? goodParse = int.tryParse('999');
  print('int.tryParse("999")    = $goodParse');

  // Type checking with is
  var value = 42;
  print('\nType checking:');
  print('  42 is int:    ${value is int}');
  print('  42 is double: ${value is double}');
  print('  42 is num:    ${value is num}');

  // String representations
  int x = 255;
  print('\nString representations of 255:');
  print('  Decimal: ${x.toString()}');
  print('  Hex:     ${x.toRadixString(16).toUpperCase()}');
  print('  Binary:  ${x.toRadixString(2)}');
}
